[ONNX] Use the stable APIs in onnxscript and sync the latest logic

justinchuby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134782

Use the stable apis from onnxscript: https://github.com/microsoft/onnxscript/issues/1827
Sync with torch-onnx at https://github.com/justinchuby/torch-onnx/compare/v0.1.12...v0.1.15.

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/exporter/_registration.py:
```diff
@@ -162,33 +159,19 @@ def from_torchlib(
+                    is_custom=False,
+                    is_complex=meta.is_complex,
+                )
+                registry._register(target, onnx_decomposition)
```
Is there still registry api for custom ops?

titaiwangms replied:
It just occurs to me that we might need to refactor ONNX tutorial on PyTorch/tutorial afer the migration.

justinchuby replied:
I agree. One thing is how to do want to expose the registration API which I don’t have a good idea yet

titaiwangms commented on torch/onnx/_internal/exporter/_compat.py:
```diff
@@ -206,20 +189,22 @@ def export_compat(
-            )
-        _convert_version(f, opset_version)
+    # Converter opset version and optimize
+    onnx_program.model = onnxscript_apis.convert_version(
```
Check if the model is already in the desired version or not?

justinchuby replied:
I have delegated this to the api itself. So the api is going to do nothing if the model version is desired

titaiwangms commented on torch/onnx/_internal/exporter/_compat.py:
```diff
@@ -206,20 +189,22 @@ def export_compat(
+    )
+    onnx_program.model = onnxscript_apis.optimize(onnx_program.model)
+
+    if f is not None:
```
f is None is deleted in main api

justinchuby replied:
I should should change that

justinchuby replied:
f can be None right? That just means the output is a nonnxprogram

justinchuby replied:
That's the same signature in `torch.onnx.__init__`

titaiwangms commented on torch/onnx/_internal/exporter/_core.py:
```diff
@@ -768,7 +766,7 @@ def _exported_program_to_onnx_program(
             },
         ),
         ir_version=9,
```
Should we hide this for version control?

justinchuby replied:
Can you elaborate?

titaiwangms replied:
I am thinking ir version should go with onnxscript version, as it's coded in onnxscript. It seems less informative when we hardcode it here. Would we have to compare ir version in debugging?

justinchuby replied:
IR version is mainly used for enabling the latest dtypes. Hard coding to 9 seems ok for this pytorch version as it has all the dtypes we need.

justinchuby commented on torch/onnx/_internal/exporter/_onnx_program.py:
```diff
@@ -5,26 +5,24 @@
 if TYPE_CHECKING:
     import onnxruntime as ort
 
+_LARGE_MODEL_THRESHOLD = 512 * 1024 * 1024  # 512MB
```
Change to 1.5gb

justinchuby committed:
Update on "[ONNX] Use the stable APIs in onnxscript"


TODO: Add the onnxscript requirement check
Sync with torch-onnx


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use the stable APIs in onnxscript and sync the latest logic"


Use the stable apis from onnxscript: https://github.com/microsoft/onnxscript/issues/1827
Sync with torch-onnx at https://github.com/justinchuby/torch-onnx/compare/v0.1.12...v0.1.15.


[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use the stable APIs in onnxscript and sync the latest logic"


Use the stable apis from onnxscript: https://github.com/microsoft/onnxscript/issues/1827
Sync with torch-onnx at https://github.com/justinchuby/torch-onnx/compare/v0.1.12...v0.1.15.


[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge -f "related tests passed"
